shared/types/Tournament.d.ts

```diff
@@ -1,27 +1,39 @@
-import User from './User';
+import User from "./User";
+import mongoose, { Schema, Document, Types } from 'mongoose';
```

Comment:
I guess that the types defined in shared are used in the "business logic". If that's correct then they should not depend on mongoose or anything related to the DB layer. A dependency from business code to infrastructural code causes many problems. One of them that comes to my mind here is "poisoning" the whole core (business) part of the application with some infrastructural quirks. IMO an implementation of a use-case should not worry about the DB stuff.

Reply:
This file only uses `mongoose.Types`, to provide this interface flexibility with help of union type
```
teams: Array<Types.ObjectId> |  Array<Team>;
team: Team | Types.ObjectId
```
A union type describes a value that can be one of several types.
I can redefine this interface (this field to be able to store Team object or ref to Team) in TournamentModel which I will do, therefore dependecy in "SharedTypes" will no longer exist.

Reply:
I know what union types are ;p I'm just wondering if we couldn't resolve references in the infrastructural (DB) layer fully and then we wouldn't have to deal with this in the business logic. Also, when it comes to `mongoose.Types.Array` and the standard `Array`, do we have to use the former for `Document` types for the whole thing to work properly? It'd be much better to not have any dependency on the DB library in these models. Even if we need to use IDs for referencing then IMO it should be some custom IDs.

Reply:
Also, if you reference mongoose in the shared model it would also carry over the dependency to the frontend module which would make no sense at all...

Reply:
The problem is that (I think) TypeScript itself, doesn't provide us with ObjectId type which we have to specify here. It has to be taken from `mongoose.Types.ObjectId`

Reply:
Why can't we use a simple type in the shared model, like a `string` or a `number` or not use an ID at all in the business layer?